prune: stop creating `package-lock.json`

watilde opened the pull request:
A package-lock.json file was created when `npm prune` is executed, and it deleted the existed dependencies/devDependencies. `npm prune` shouldn not touch `package-lock.json` from the first.

related issues:
+ https://github.com/npm/npm/issues/17061
+ https://github.com/npm/npm/issues/17068

watson commented on lib/prune.js:
```diff
@@ -21,6 +21,8 @@ prune.completion = require('./utils/completion/installed-deep.js')
 function prune (args, cb) {
   var dryrun = !!npm.config.get('dry-run')
+  // prune should not touch the lock file
+  npm.config.set('save', false)
```
I think the proper config option to use for this is `shrinkwrap`. The package-lock.json file is only not produced as a side-effect of setting `save` to `false`.

watson replied:
Btw, isn't setting this config option a permanent thing? Or is it just for this command and this run?

watilde replied:
I think `npm.config.set('shrinkwrap', false)` does not work because we need to avoid `saveToDependencies` at [lib/install.js#L294-L300](https://github.com/npm/npm/blob/90a25f65804d7d7dc351d01c53826b879139d8da/lib/install.js#L294-L300) to not make the lock file with `npm prune`, and the flag comes from here [lib/install/save.js#L133-L136](https://github.com/npm/npm/blob/latest/lib/install/save.js#L133-L136). The default value of the save option is `true` since npm@5, then we can turn off it at the prune class.

`npm.config.set` is to set a value to a config property in runtime. It doesn't overwrite `.npmrc`.

watson replied:
Ah ok, that sounds right then 👍

iarna commented:
I'm not sure this is right… if `npm prune` is modifying the tree then it needs to update any existing `package-lock` or `shrinkwrap`. I think what you want is for it to not create one if one isn't already there, which will require some new instrumenting over in `save` I think.

iarna reviewed (changes requested):
I'm not sure this is right… if `npm prune` is modifying the tree then it needs to update any existing `package-lock` or `shrinkwrap`. I think what you want is for it to not create one if one isn't already there, which will require some new instrumenting over in `save` I think.

watilde commented:
That makes sense! Then I will update this process to:
+ `npm prune` doesn't make `package-lock.json` when it doesn't exist.

// correct me if I'm misunderstanding :)

watilde committed:
prune: don't create `package-lock.json`

A package-lock.json file was created `npm prune` is executed,
and it deleted the existed dependencies/devDependencies.
`npm prune` shouldn't touch `package-lock.json` from the first.

watilde commented:
I've updated as I commented.

iarna commented:
@watilde Forgive me only asking this now, but why shouldn't prune create a `package-lock.json`? If you don't want one, wouldn't you be running with `package-lock = false` in your config?

jasonraimondi commented:
Running `npm prune --production` is updating my `package-lock.json` file on a prod deploy. It would be great if there was a way to flag `--no-lockfile` on a prune run to get this to stop updating.

Currently I am needing to reset lock changes on the step after the prune to stop the lockfile changes: `git checkout package-lock.json`
